File: src/main/java/com/example/jwttest/service/UserService.java
package com.example.jwttest.service;

import com.example.jwttest.exception.ErrorCode;
import com.example.jwttest.dto.UserDto;
import com.example.jwttest.entity.Authority;
import com.example.jwttest.entity.RefreshToken;
import com.example.jwttest.entity.User;
import com.example.jwttest.exception.CustomException;
import com.example.jwttest.repository.UserRepository;
import com.example.jwttest.util.SecurityUtil;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collections;
import java.util.UUID;

/**
 * 회원가입, 유저정보 조회
 */
@Service
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public UserService(UserRepository userRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    /**
     * 회원가입
     */
    @Transactional
    public UserDto signup(UserDto userDto) {
        if (userRepository.findOneWithAuthoritiesByUsername(userDto.getUsername()).orElse(null) != null) {
            throw new CustomException(ErrorCode.DUPLICATE_USER);
        }

        Authority authority = Authority.builder()
                .authorityName("ROLE_USER")                         // 회원가입을 통한 유저는 권한이 USER
                .build();

        User user = User.builder()
                .username(userDto.getUsername())
                .password(passwordEncoder.encode(userDto.getPassword()))
                .nickname(userDto.getNickname())
                .authorities(Collections.singleton(authority))
                .activated(true)
                .build();

        return UserDto.from(userRepository.save(user));
    }

    /**
     * 내 정보 조회 (ROLE_USER)
     */
    @Transactional(readOnly = true)
    public UserDto getUserWithAuthorities(String username) {
//        return UserDto.from(userRepository.findOneWithAuthoritiesByUsername(username).orElse(null));
        return UserDto.from(
                userRepository.findOneWithAuthoritiesByUsername(username)
                        .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND))
        );
    }

    /**
     * 유저정보 조회 (ROLE_ADMIN)
     */
    @Transactional(readOnly = true)
    public UserDto getMyUserWithAuthorities() {
        return UserDto.from(
                SecurityUtil.getCurrentUsername()                                       // SecurityContext 에서 username 을 가져온다.
                        .flatMap(userRepository::findOneWithAuthoritiesByUsername)      // username 을 기준으로 User 정보를 가져온다.
                        .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND))
        );
    }



}

File: src/main/kotlin/com/example/jwttest/service/UserService.kt
package com.example.jwttest.service

import com.example.jwttest.dto.UserDto
import com.example.jwttest.dto.UserDto.Companion.from
import com.example.jwttest.entity.Authority
import com.example.jwttest.entity.User
import com.example.jwttest.exception.CustomException
import com.example.jwttest.exception.ErrorCode
import com.example.jwttest.repository.UserRepository
import com.example.jwttest.util.SecurityUtil
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * 회원가입, 유저정보 조회
 */
@Service
class UserService(private val userRepository: UserRepository, private val passwordEncoder: PasswordEncoder) {
    /**
     * 회원가입
     */
    @Transactional
    fun signup(userDto: UserDto?): UserDto? {
        if (userRepository.findOneWithAuthoritiesByUsername(userDto?.username!!).orElse(null) != null) {
            throw CustomException(ErrorCode.DUPLICATE_USER)
        }
        val authority = Authority(authorityName = "ROLE_USER")
        val user = User(
            username = userDto.username,
            password = passwordEncoder.encode(userDto.password),
            nickname = userDto.nickname,
            authorities = setOf(authority),
            isActivated = true
        )

        return from(userRepository.save(user))
    }

    /**
     * 내 정보 조회 (ROLE_USER)
     */
    @Transactional(readOnly = true)
    fun getUserWithAuthorities(username: String): UserDto {
        return from(
            userRepository.findOneWithAuthoritiesByUsername(username)
                .orElse(null)
        )
    }

    /**
     * 유저정보 조회 (ROLE_ADMIN)
     */
    @get:Transactional(readOnly = true)
    val myUserWithAuthorities: UserDto
        get() = from(
            SecurityUtil.currentUsername                                                                                // SecurityContext 에서 username 을 가져온다.
                .flatMap { username: String -> userRepository.findOneWithAuthoritiesByUsername(username)  }            // username 을 기준으로 User 정보를 가져온다.
                .orElseThrow { CustomException(ErrorCode.USER_NOT_FOUND) }
        )
}